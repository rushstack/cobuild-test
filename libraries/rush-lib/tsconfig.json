{
  "extends": "./node_modules/@rushstack/heft-node-rig/profiles/default/tsconfig-base.json",
  "compilerOptions": {
    "types": ["heft-jest", "node", "webpack-env"],
    "skipLibCheck": true,
    "resolveJsonModule": true,
    "outDir": "lib-commonjs"
  }
}
