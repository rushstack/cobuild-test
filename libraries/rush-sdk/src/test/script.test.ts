import * as path from 'path';
import { Executable } from '@rushstack/node-core-library';

const rushSdkPath: string = path.join(__dirname, '../../lib-shim/index.js');
const sandboxRepoPath: string = `${__dirname}/sandbox`;
const mockPackageFolder: string = `${sandboxRepoPath}/mock-package`;
const mockRushLibPath: string = `${__dirname}/fixture/mock-rush-lib.js`;

const coreLibPath: string = require.resolve('@rushstack/node-core-library');

describe('@rushstack/rush-sdk', () => {
  it('Should load via global (for plugins)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
global.___rush___rushLibModule = { foo: 1 };
console.log(Object.keys(require(${JSON.stringify(rushSdkPath)})));`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via env when Rush has loaded (for child processes)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
require('@microsoft/rush-lib');
console.log(Object.keys(require(${JSON.stringify(rushSdkPath)})));`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via process.env._RUSH_LIB_PATH (for child processes)', () => {
    const result = Executable.spawnSync(
      'node',
      ['-e', `console.log(Object.keys(require(${JSON.stringify(rushSdkPath)})));`],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: mockRushLibPath
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via install-run (for standalone tools)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
const { Import } = require(${JSON.stringify(coreLibPath)});
const originalResolveModule = Import.resolveModule;
const mockResolveModule = (options) => {
  if (options.baseFolderPath.includes('install-run') && options.modulePath === '@microsoft/rush-lib') {
    return ${JSON.stringify(mockRushLibPath)};
  }
  return originalResolveModule(options);
}
Import.resolveModule = mockResolveModule;
console.log(Object.keys(require(${JSON.stringify(rushSdkPath)})));
`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });
});
