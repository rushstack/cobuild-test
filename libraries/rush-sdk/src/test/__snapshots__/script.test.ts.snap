// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@rushstack/rush-sdk Should load via env when Rush has loaded (for child processes): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via env when Rush has loaded (for child processes): stdout 1`] = `
"Try to load @microsoft/rush-lib from process.env._RUSH_LIB_PATH from caller package
Loaded @microsoft/rush-lib from process.env._RUSH_LIB_PATH
[
  '_rushSdk_loadInternalModule',
  'ApprovedPackagesConfiguration',
  'ApprovedPackagesItem',
  'ApprovedPackagesPolicy',
  'BuildCacheConfiguration',
  'BumpType',
  'ChangeManager',
  'CobuildConfiguration',
  'CommonVersionsConfiguration',
  'CredentialCache',
  'CustomTipsConfiguration',
  'DependencyType',
  'EnvironmentConfiguration',
  'EnvironmentVariableNames',
  'Event',
  'EventHooks',
  'ExperimentsConfiguration',
  'FileSystemBuildCacheProvider',
  'IndividualVersionPolicy',
  'LockStepVersionPolicy',
  'LookupByPath',
  'NpmOptionsConfiguration',
  'Operation',
  'OperationStatus',
  'PackageJsonDependency',
  'PackageJsonEditor',
  'PackageManager',
  'PackageManagerOptionsConfigurationBase',
  'PhasedCommandHooks',
  'PnpmOptionsConfiguration',
  'ProjectChangeAnalyzer',
  'RepoStateFile',
  'Rush',
  'RushConfiguration',
  'RushConfigurationProject',
  'RushConstants',
  'RushLifecycleHooks',
  'RushProjectConfiguration',
  'RushSession',
  'RushUserConfiguration',
  'VersionPolicy',
  'VersionPolicyConfiguration',
  'VersionPolicyDefinitionName',
  'YarnOptionsConfiguration',
  '_LastInstallFlag',
  '_OperationMetadataManager',
  '_OperationStateFile',
  '_RushGlobalFolder',
  '_RushInternals'
]"
`;

exports[`@rushstack/rush-sdk Should load via global (for plugins): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via global (for plugins): stdout 1`] = `"[ '_rushSdk_loadInternalModule', 'foo' ]"`;

exports[`@rushstack/rush-sdk Should load via install-run (for standalone tools): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via install-run (for standalone tools): stdout 1`] = `
"Trying to load  @microsoft/rush-lib installed by install-run-rush
Loaded @microsoft/rush-lib installed by install-run-rush
[ '_rushSdk_loadInternalModule', 'foo' ]"
`;

exports[`@rushstack/rush-sdk Should load via process.env._RUSH_LIB_PATH (for child processes): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via process.env._RUSH_LIB_PATH (for child processes): stdout 1`] = `
"Try to load @microsoft/rush-lib from process.env._RUSH_LIB_PATH from caller package
Loaded @microsoft/rush-lib from process.env._RUSH_LIB_PATH
[ '_rushSdk_loadInternalModule', 'foo' ]"
`;
